zebra/tc_netlink.c

```diff
@@ -364,105 +522,361 @@ static ssize_t netlink_tfilter_msg_encode(int cmd, struct zebra_dplane_ctx *ctx,
+					      void *buf, size_t buflen)
+{
+	return netlink_tfilter_msg_encode(RTM_NEWTFILTER, ctx, buf, buflen);
+}
 
-	nl_attr_put(&req->n, datalen, TCA_KIND, kind, strlen(kind) + 1);
-	nest = nl_attr_nest(&req->n, datalen, TCA_OPTIONS);
+static ssize_t netlink_deltfilter_msg_encoder(struct zebra_dplane_ctx *ctx,
+					      void *buf, size_t buflen)
+{
+	return netlink_tfilter_msg_encode(RTM_DELTFILTER, ctx, buf, buflen);
+}
 
-	nl_attr_put(&req->n, datalen, TCA_FLOWER_CLASSID, &classid,
-		    sizeof(classid));
+enum netlink_msg_status
+netlink_put_tc_qdisc_update_msg(struct nl_batch *bth,
+				struct zebra_dplane_ctx *ctx)
+{
+	enum dplane_op_e op;
+	enum netlink_msg_status ret;
 
-	if (filter_bm & TC_FILTER_SRC_IP) {
-		const struct prefix *src_p = dplane_ctx_tc_get_src_ip(ctx);
+	op = dplane_ctx_get_op(ctx);
 
-		if (tc_flower_get_inet_prefix(src_p, &addr) != 0)
-			return 0;
+	if (op == DPLANE_OP_TC_QDISC_INSTALL) {
+		ret = netlink_batch_add_msg(
+			bth, ctx, netlink_newqdisc_msg_encoder, false);
+	} else if (op == DPLANE_OP_TC_QDISC_UNINSTALL) {
+		ret = netlink_batch_add_msg(
+			bth, ctx, netlink_delqdisc_msg_encoder, false);
+	} else {
+		return FRR_NETLINK_ERROR;
+	}
 
-		nl_attr_put(&req->n, datalen,
-			    (addr.family == AF_INET) ? TCA_FLOWER_KEY_IPV4_SRC
-						     : TCA_FLOWER_KEY_IPV6_SRC,
-			    addr.data, addr.bytelen);
+	return ret;
+}
 
-		if (tc_flower_get_inet_mask(src_p, &addr) != 0)
-			return 0;
+enum netlink_msg_status
+netlink_put_tc_class_update_msg(struct nl_batch *bth,
+				struct zebra_dplane_ctx *ctx)
+{
+	enum dplane_op_e op;
+	enum netlink_msg_status ret;
 
-		nl_attr_put(&req->n, datalen,
-			    (addr.family == AF_INET)
-				    ? TCA_FLOWER_KEY_IPV4_SRC_MASK
-				    : TCA_FLOWER_KEY_IPV6_SRC_MASK,
-			    addr.data, addr.bytelen);
+	op = dplane_ctx_get_op(ctx);
+
+	if (op == DPLANE_OP_TC_CLASS_ADD || op == DPLANE_OP_TC_CLASS_ADD) {
+		ret = netlink_batch_add_msg(
+			bth, ctx, netlink_newtclass_msg_encoder, false);
+	} else if (op == DPLANE_OP_TC_CLASS_DELETE) {
+		ret = netlink_batch_add_msg(
+			bth, ctx, netlink_deltclass_msg_encoder, false);
+	} else {
+		return FRR_NETLINK_ERROR;
 	}
 
-	if (filter_bm & TC_FILTER_DST_IP) {
-		const struct prefix *dst_p = dplane_ctx_tc_get_dst_ip(ctx);
+	return ret;
+}
 
-		if (tc_flower_get_inet_prefix(dst_p, &addr) != 0)
-			return 0;
+enum netlink_msg_status
+netlink_put_tc_filter_update_msg(struct nl_batch *bth,
+				 struct zebra_dplane_ctx *ctx)
+{
+	enum dplane_op_e op;
+	enum netlink_msg_status ret;
+
+	op = dplane_ctx_get_op(ctx);
+
+	if (op == DPLANE_OP_TC_FILTER_ADD) {
+		ret = netlink_batch_add_msg(
+			bth, ctx, netlink_newtfilter_msg_encoder, false);
+	} else if (op == DPLANE_OP_TC_FILTER_UPDATE) {
+		/*
+		 * Replace will fail if either filter type or the number of
+		 * filter options is changed, so DEL then NEW
+		 *
+		 * TFILTER may have refs to TCLASS.
+		 */
+
+		ret = netlink_batch_add_msg(
```

Comment:
will it look fine if we ignore the first `ret`? since static analyzer is complaining this.